docs/cip/CIP-10112023_Authorization.md

```diff
@@ -0,0 +1,148 @@
+
+Authorization response provides authorization provider evaluation response. It returns a boolean response indicating whether the request is allowed or denied.
+
+#### ChromaAuthzMiddleware
+
+The `ChromaAuthzMiddleware` is an abstraction for the server-side middleware. At the time of writing we only support FastAPI. The  middleware interface supports several methods:
+
+- `authorize` - authorizes the request against the authorization provider.
+- `ignore_operation` - determines whether or not the operation should be ignored by the middleware
+- `instrument_server` - an optional method for additional server instrumentation. For example, header injection.
+
+#### AuthorizationError
+
+Error thrown when an authorization request is disallowed/denied by the authorization provider. Depending on authorization provider's implementation such error may also be thrown when the authorization provider is not available or an internal error ocurred.
+
+Client semantics of this error is a 403 Unauthorized error being returned over HTTP interface.
+
+#### AuthorizationContext
+
+The AuthorizationContext provides the necessary information to the authorization provider to make a decision whether to allow or deny the request. The context contains the following information:
+
+```json
+{
+"user": {"id": "API Token or User Id"},
+"resource": {"namespace": "database", "id": "collection_id"},
+"action": {"id":"get_or_create"},
+}
+```
+
+We intentionally want to keep this as minimal as possible to avoid any unnecessary complexity and to allow users to easily understand the authorization model. However the context is just an abstraction of the above representation and each authorization provider will need to implement the above and if necessary extend it to support additional information.
+
+We propose the following classes to represent the above:
+
+```python
+from typing import TypeVar, Generic
+
+# Below three classes are from Chroma's baseline AuthZ implementation
+class User:
+    id: Optional[str]
+
+class Resource:
+    id: str
+    namespace: Optional[str]
+
+class Action:
+    id: str
+
+TUser = TypeVar("TUser")
+TResource = TypeVar("TResource")
+TAction = TypeVar("TAction")
+
+class AbstractAuthZContext(Generic[TUser, TResource, TAction]):
+    user: TUser
+    resource: TResource
+    action: TAction
+
+# Example of a concrete implementation from Chroma's baseline AuthZ implementation
+
+class SimpleAuthZContext(AbstractAuthZContext[User, Resource, Action]):
+    pass
+```
+
+### Baseline Implementation
+
+In this section we propose a minimal implementation example of the authorization framework which will also ship in Chroma as a default authorization provider and a reference implementation. Our reference implementation relies on static configuration files in YAML format.
+
+#### Authentication and Authorization Config Scheme
+
+```yaml
+users:
+    - id: user@example.com
+      tokens:
+        - my_api_token
+actions:
+    - list_collections
```

Comment:
does this simple scheme just grant these for all collections? that makes sense - just want to understand

Reply:
I'm reworking this to be more aligned with what's out there:

Here's how roles and resource type to action mappings look like:

```yaml
resource_type_action:
    - db:list_collections
    - db:get_collection
    - db:create_collection
    - db:get_or_create_collection
    - db:delete_collection
    - db:update_collection
    - collection:add
    - collection:delete
    - collection:get
    - collection:query
    - collection:peek
    - collection:update
    - collection:upsert

roles_mapping:
    - id: admin
      actions: *
    - id: write
      actions: [db:list_collections, db:get_collection, db:create_collection, db:get_or_create_collection, db:delete_collection, db:update_collection, collection:add, collection:delete, collection:get, collection:query, collection:peek, collection:update, collection:upsert]
    - id: db_read
      actions: [db:list_collections, db:get_collection, db:create_collection, db:get_or_create_collection, db:delete_collection, db:update_collection]
    - id: collection_read
      actions: [db:list_collections, db:get_collection,collection:get, collection:query, collection:peek]
    - id: collection_read
      actions: [db:get_collection,collection:get, collection:query, collection:peek]
      resources: ["<UUID>"]
```

The above is the simplest way to implement roles and actions that roles can perform. We also provide mechanism to restrict roles to specific resource ids (for now this is just collection uuids)

and for the users:

```yaml
-   id: user@example.com
    role: admin
    tokens:
        -   token: my_api_token
            secret: my_api_secret
```

The token part can work with and without the secret part, but this requires a new authentication provider to verify tokens and inject user identity (in this case, just the id) into the request context so that authz can pick it up. Alternatively, I also think we can do it in one go where the authz provider takes care of matching the token to the user id and then match the roles (this is similar to what Auth0 are doing with their machine-to-machine flow)